pennylane/ops/op_math/evolve.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-few-public-methods,function-redefined
+
+"""
+This file contains the ``Evolve`` operator.
+"""
+
+from typing import Union
+
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation, Operator
+
+from .exp import Evolution
+from .parametrized_hamiltonian import ParametrizedHamiltonian
+
+
+def evolve(op: Union[Operator, ParametrizedHamiltonian]):
+    """Returns a new operator that computes the evolution of the given operator.
+
+    Args:
+        op (Operator): operator to evolve
+
+    Returns:
+        Evolution | ParametrizedEvolution: evolution operator
+    """
+    if isinstance(op, ParametrizedHamiltonian):
+
+        def parametrized_evolution(params: list, t, dt=None):
+            """Constructor for the :class:`ParametrizedEvolution` operator.
+
+            Args:
+                params (list): params (ndarray): trainable parameters
+                t (Union[float, List[float]]): If a float, it corresponds to the duration of the evolution.
+                    If a list of two floats, it corresponds to the initial time and the final time of the
+                    evolution.
+                dt (float, optional): the time step used by the differential equation solver to evolve the
+                    time-dependent Hamiltonian. Defaults to XXX.
+
+            Returns:
+                ParametrizedEvolution: class used to compute the parametrized evolution of the given
+                    hamiltonian
+            """
+            return Evolve(H=op, params=params, t=t, dt=dt)
+
+        return parametrized_evolution
```

Comment:
So the various options here are:
1. Return a function, `parametrized_evolution` that gives an `Evolve` when called (i.e., what's done now)
2. Return an `Evolve` operation, which has a call method that fills in its additional parameters (`params` and `t`)

One drawback of option 2 is that it might not make sense to have a "partial" operation (i.e., before the `params` and `t` are fixed). Though one advantage of option 2 is ease of inspection - users can see the `H` to be exponentiated for example.

What was the motivation for the choice of option 1 over option 2?